DRAFT!!!!!!11 schema_applier: unify handling of token_metadata during schema change

nuivall opened the pull request:
TODO:
- cover letter
- split commit_tablet_metadata into prepare + commit phases
- more testing runs

Fixes #https://github.com/scylladb/scylladb/issues/24414

ptrsmrn commented on db/schema_applier.cc:
```diff
@@ -466,6 +466,8 @@ future<> schema_applier::merge_types()
         af.dropped = co_await create_types(db, diff.dropped, new_keyspaces_per_shard);
     });
+
+    co_await _types_storage.init(_proxy.local().get_db(), _affected_keyspaces, _affected_user_types);
```
But why move it?

nuivall replied:
added

avikivity commented on service/migration_manager.cc:
```diff
@@ -112,8 +112,15 @@ future<> migration_manager::drain()
+            co_await reload_schema().handle_exception([] (std::exception_ptr ep) {
                 // Due to features being unordered, reload might fail because
                 // some tables still have the wrong version and looking up e.g.
                 // the base-table of a view will fail.
```
Since you coroutinized this, also fix  up the handle_exception. Mixed continuations and coroutines are harder to understand.

Do we need coroutine::lambda() here? The lambda coroutine tranlates to coroutine_func(lambda_captures_holding_this&), which in turn means the lambda_captures_holding_this is not copied to the coroutine frame, only a reference to it.

Safer to pass this as a parameter and copy it to the coroutine body.

Even better is to make it a fiber in migration_manager, signal it when it needs to work, and wait for it in migration_manager::stop().

avikivity commented on service/storage_service.hh:
```diff
@@ -148,6 +149,15 @@ struct token_metadata_change {
+    virtual replica::column_family* find_column_family(const table_id& uuid) const = 0;
+    virtual flat_hash_map<sstring, replica::keyspace*> get_keyspaces() const = 0;
+    virtual future<> for_each_table_gently(std::function<future<>(table_id, lw_shared_ptr<replica::table>)> f) const = 0;
+};
```
schema_getter smells to much of data_dictionary. Is this about the schema, or the replica side?

nuivall replied:
virtual destructor not needed as I only pass base class via reference, but agree it's safer to add, so I'll add it.

 schema_getter is very similar to data_dictionary::database, but it's pure virtual interface and contains only a tiny subset needed for `prepare_token_metadata_change`. Otherwise I'd need to implement much more functions.

But I can move it to data_dictionary namespace, is that your suggestion?

nuivall replied:
after deeper look I can't move it to data_dictionary as it's not using those "shell" `table`, `keyspace` objects defined there but real ones defined in replica/database.hh

avikivity commented on db/schema_applier.cc:
```diff
@@ -841,6 +842,79 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+        return &_db.local().find_column_family(uuid);
+    };
+
+    virtual flat_hash_map<sstring, replica::keyspace*> get_keyspaces() const {
```
(have my own flattening of storage_proxy in prep)

nuivall replied:
in theory it's used after merge_keyspaces so it can be pre-computed
- add something like `future<> pending_schema_getter::prepare`
- keep `pending_schema_getter` object in schema_applier
- make `get_keyspaces` return shared pointer to flat_hash_map

should I do it?

bitpathfinder replied:
It should be (16+8)*5000 bytes = 120kb (plus some flat hashmap data). Will it trigger a warning?

nuivall commented on service/migration_manager.cc:
```diff
@@ -112,8 +112,15 @@ future<> migration_manager::drain()
+                if (_as.abort_requested()) {
+                    co_return;
+                }
+                co_await seastar::sleep_abortable(10ms, _as);
```
no impact as this is done  in the background

also btw as I understand https://github.com/scylladb/scylladb/pull/16013 it's probably not needed in raft mode

nuivall commented on service/storage_service.hh:
```diff
@@ -138,6 +138,16 @@ class node_ops_meta_data;
+    std::vector<std::unordered_map<table_id, locator::effective_replication_map_ptr>> pending_view_erms{smp::count};
+    std::unordered_set<session_id> open_sessions;
+
+    future<> destroy();
```
ok

nuivall replied:
yes, but is this much better? I tried to stick to the original code, doing as little intermediate refactor as possible

nuivall commented:
v2:
- changed reload_schema().handle_exception to try/catch
- added virtual destrutor in schema_getter

nuivall commented on service/storage_service.cc:
```diff
@@ -3289,9 +3293,32 @@ void storage_service::commit_token_metadata_change(token_metadata_change& change
+        virtual replica::column_family* find_column_family(const table_id& uuid) const {
+            return &_db.local().find_column_family(uuid);
+        };
+        virtual flat_hash_map<sstring, replica::keyspace*> get_keyspaces() const {
```
due to `class pending_schema_getter` implementation

nuivall commented on service/storage_service.cc:
```diff
@@ -3150,34 +3150,33 @@ future<token_metadata_change> storage_service::prepare_token_metadata_change(mut
-            auto& db = ss._db.local();
-            for (auto& ks_name : keyspaces) {
-                auto rs = db.find_keyspace(ks_name).get_replication_strategy_ptr();
+            auto keyspaces = schema_getter.get_keyspaces();
```
it's pre-existing code, I tried to not do any additional refactor, not needed for the split here

nuivall commented on service/storage_service.cc:
```diff
@@ -3289,9 +3293,32 @@ void storage_service::commit_token_metadata_change(token_metadata_change& change
+    public:
+        db_schema_getter(distributed<replica::database>& db) : _db(db) {};
+
+        virtual replica::column_family* find_column_family(const table_id& uuid) const {
```
I put a comment in class schema_getter interface:
    // find_column_family returns nullptr only when table was deleted
    // during pending schema change.

nuivall commented on service/migration_manager.cc:
```diff
@@ -112,13 +113,22 @@ future<> migration_manager::drain()
+                if (_as.abort_requested()) {
+                    co_return;
+                }
+                co_await seastar::sleep_abortable(10ms, _as);
```
yes, good point

nuivall commented:
v3:
- fixed indent
- changed abort_requested check to try catch sleep_aborted

andrzej-jackowski-scylladb commented on db/schema_applier.cc:
```diff
@@ -981,8 +985,8 @@ future<> schema_applier::finalize_tables_and_views() {
     // TODO: maybe untangle this dependency
-    if (diff.new_token_metadata.local()) {
-        co_await _ss.local().commit_tablet_metadata(diff.new_token_metadata.local());
+    if (_tablet_hint) {
```
If you want to change `diff.new_token_metadata.local()` to `_tablet_hint` maybe do it in a separate commit with a description

andrzej-jackowski-scylladb commented on db/schema_applier.cc:
```diff
@@ -852,12 +845,23 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+// Loads metadata into a single source of truth to ensure consistency.
+// It will be committed later if required by the current schema change.
+future<> schema_applier::load_mutable_token_metadata() {
+    locator::mutable_token_metadata_ptr new_token_metadata = nullptr;
```
We don't need current_token_metadata in the outer scope

nuivall replied:
will move

andrzej-jackowski-scylladb commented on service/storage_service.cc:
```diff
@@ -3150,34 +3150,33 @@ future<token_metadata_change> storage_service::prepare_token_metadata_change(mut
-                auto rs = db.find_keyspace(table->schema()->ks_name()).get_replication_strategy_ptr();
+            auto keyspaces = schema_getter.get_keyspaces();
+            co_await schema_getter.for_each_table_gently([&] (table_id id, lw_shared_ptr<replica::table> table) {
+                auto rs = keyspaces.at(table->schema()->ks_name())->get_replication_strategy_ptr();
```
We change the exception type that we are throwing from `no_such_keyspace` to `std::out_of_range`. Please verify it's not a problem or make the exception type the same as earlier.

andrzej-jackowski-scylladb commented on service/storage_service.cc:
```diff
@@ -3289,9 +3293,32 @@ void storage_service::commit_token_metadata_change(token_metadata_change& change
+            }
+            return out;
+        };
+        virtual future<> for_each_table_gently(std::function<future<>(table_id, lw_shared_ptr<replica::table>)> f) const {
```
Please add override and also consider removing virtual.

andrzej-jackowski-scylladb commented on db/schema_applier.cc:
```diff
@@ -841,6 +842,79 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+            _sa(sa), _db(sa._proxy.local().get_db()) {
+    };
+
+    virtual replica::column_family* find_column_family(const table_id& uuid) const {
```
Consider using override

nuivall commented on service/storage_service.cc:
```diff
@@ -3289,9 +3293,32 @@ void storage_service::commit_token_metadata_change(token_metadata_change& change
+    public:
+        db_schema_getter(distributed<replica::database>& db) : _db(db) {};
+
+        virtual replica::column_family* find_column_family(const table_id& uuid) const {
```
will add override, but we don't remove virtual in such cases

nuivall commented:
@lsfreitas you really need to improve Jenkins<>Github integration, it's not good to report green here and failed on jenkins, and it's happening notoriously. But anyway the failure is unrelated.

nuivall commented:
rebased to re-trigger CI

lsfreitas commented:
> @lsfreitas you really need to improve Jenkins<>Github integration, it's not good to report green here and failed on jenkins, and it's happening notoriously. But anyway the failure is unrelated.

It shouldn't be happening anymore. There was a bug on the implementation that was reverted on the day I set the comment. It should be stable by now

tgrabiec commented on db/schema_applier.cc:
```diff
@@ -845,6 +841,19 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+    auto& db = _proxy.local().get_db();
+    _metadata_locks = std::make_unique<
+            replica::tables_metadata_lock_on_all_shards>(
+            co_await replica::database::lock_tables_metadata(db));
```
Why not take it in commit() on each shard to reduce the scope of the lock and thus impact on concurrent operations?

tgrabiec commented on db/schema_applier.cc:
```diff
@@ -841,6 +842,79 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+                  co_await f(id, t);
+            }
+        };
+        return _db.local().get_tables_metadata().for_each_table_gently(ff);
```
What about new tables?

nuivall replied:
but it's problematic because at this point there is no table object yet, and because committing tables uses _pending_token_metadata which is up to date, the new change is included there.

tgrabiec commented on db/schema_applier.cc:
```diff
@@ -912,6 +986,15 @@ void schema_applier::commit_tables_and_views() {
+    // commit token_metadata, must be before updating keyspaces as
+    // both may update some metadata (namely erm) but only keyspace operations
+    // include up-to-date information (because they combine keyspace change and
+    // tablet change - via pending_token_metadata).
```
This sounds odd. Can you elaborate what metadata is updated later by keyspace operations?

nuivall replied:
commit_token_metadata_change calls ks->update_static_effective_replication_map which overwrites _effective_replication_map

keyspace::apply is also changing _effective_replication_map

tgrabiec commented on db/schema_applier.cc:
```diff
@@ -841,6 +842,79 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+            out.insert_or_assign(name, &ks);
+        }
+        // Note that this may return stale information for altered keyspaces
+        // but we can't easily return replica::keyspace* in such case because
```
Another argument for moving commit_token_metadata_change() after schema commit.

tgrabiec commented on service/storage_service.cc:
```diff
@@ -5663,7 +5663,7 @@ future<> storage_service::keyspace_changed(const sstring& ks_name) {
 }
 
-future<locator::mutable_token_metadata_ptr> storage_service::prepare_tablet_metadata(const locator::tablet_metadata_change_hint& hint, mutable_token_metadata_ptr pending_token_metadata) {
+future<locator::mutable_token_metadata_ptr> storage_service::prepare_token_metadata_with_tablets_change(const locator::tablet_metadata_change_hint& hint, mutable_token_metadata_ptr pending_token_metadata) {
```
To elaborate, an operation on the whole doesn't have to say it's working on the whole. All instance methods are like that, e.g. this->prepare_tables_and_views.  prepare_tables_and_views(this) is the same, it's just convention that the object is not passed as parameter. token_metadata is the same.

nuivall commented on db/schema_applier.cc:
```diff
@@ -852,12 +845,23 @@ future<> schema_applier::prepare(utils::chunked_vector<mutation>& muts) {
+    locator::mutable_token_metadata_ptr current_token_metadata = co_await _ss.local().get_mutable_token_metadata_ptr();
+    if (_tablet_hint) {
+        slogger.info("Tablet metadata changed");
+        new_token_metadata = co_await _ss.local().prepare_tablet_metadata(
```
isn't this about some previous version of the PR? Currently there is auto here: https://github.com/scylladb/scylladb/pull/25302/commits/b77730f454a18f139b9c62760aecf2323a6b13e7#diff-7dbbc553212d1fb3c6dc48d3476d67d9a74a416e7f0cb7cac90042ddeb7e569bR854

nuivall commented on service/storage_service.cc:
```diff
@@ -3329,6 +3318,27 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
+    return smp::invoke_on_all([this] () {
+        pending_token_metadata_ptr[this_shard_id()] = nullptr;
+        pending_effective_replication_maps[this_shard_id()] = {};
+        pending_table_erms[this_shard_id()] = {};
```
ok

nuivall commented on service/storage_service.cc:
```diff
@@ -3261,41 +3260,46 @@ future<token_metadata_change> storage_service::prepare_token_metadata_change(mut
+                    // During pending schema change view may be deleted
+                    // but cf->views returns it as change is not commited yet
+                    // so we check with schema_getter which takes this into account.
+                    continue;
```
I think due to:
```
    // commit token_metadata, must be before updating keyspaces as
    // both may update some metadata (namely erm) but only keyspace operations
    // include up-to-date information (because they combine keyspace change and
    // tablet change - via pending_token_metadata).
```

nuivall commented:
v5:
- split new_token_metadata refactor to 2 commits
- slight rework of reload_schema_in_bg to not call it on boot but call it on feature change
- moved get_mutable_token_metadata_ptr from prepare_tablet_metadata body to argument
- added utils::clear_gently in destroy
- moved _metadata_locks to commit, slight decrese in lock scope

nuivall commented:
v6:
- schema_getter - uses get_keyspaces_replication instead of get_keyspaces_replication which allows it to accurately return pending replications
- schema_getter - for_each_table_gently changed to work on schema not replica:table so it can fully support pending changes
- commit_on_shard - commit_token_metadata_change may be called in the same order as during preparation as now it doesn't have order dependency on keyspace commit
- commit_token_metadata_change - doesn't use schema_getter anymore, it uses db as before

tgrabiec commented on service/storage_service.cc:
```diff
@@ -3270,10 +3267,10 @@ future<token_metadata_change> storage_service::prepare_token_metadata_change(mut
-            co_await db.get_tables_metadata().for_each_table_gently([&] (table_id id, lw_shared_ptr<replica::table> table) {
-                auto rs = db.find_keyspace(table->schema()->ks_name()).get_replication_strategy_ptr();
+            auto keyspaces = schema_getter.get_keyspaces();
+            co_await schema_getter.for_each_table_gently([&] (table_id id, lw_shared_ptr<replica::table> table) {
```
bisectability: There is no schema_getter::for_each_table_gently() in this commit.
